Packs/PAN-OS/Integrations/Panorama/Panorama.py

```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+        This list will only retrieve devices that are connected to panorama.
+        """
+        ha_pair_dict = {}
+        device_op_command_result: Element = run_op_command(device, "show devices all")
+        device_entry: Element
+        for device_entry in device_op_command_result.findall("./result/devices/entry"):
+            serial_number: str = find_text_in_element(device_entry, "./serial")
+            connected: str = find_text_in_element(device_entry, "./connected")
+            if connected == "yes":
+                new_firewall_object = Firewall(serial=serial_number)
+                device.add(new_firewall_object)
+                self.add_device_object(new_firewall_object)
+                ha_peer_serial_element: Optional[Element] = device_entry.find("./ha/peer/serial")
+                ha_peer_serial = None
+                if ha_peer_serial_element and hasattr(ha_peer_serial_element, "text"):
+                    ha_peer_serial = ha_peer_serial_element.text
+
+                if ha_peer_serial:
+                    # The key is always the active device.
+                    ha_status: str = find_text_in_element(device_entry, "./ha/state")
+                    if ha_status == "active":
+                        self.ha_active_devices[serial_number] = ha_peer_serial
+
+                    ha_pair_dict[serial_number] = ha_peer_serial
+                else:
+                    self.ha_active_devices[serial_number] = "STANDALONE"
+
+        self.ha_pair_serials = ha_pair_dict
+
+    def add_device_object(self, device: Union[PanDevice, Panorama, Firewall]):
+        if type(device) is Panorama:
+            # Check if HA is active and if so, what the system state is.
+            panorama_ha_state_result: Element = run_op_command(device, "show high-availability state")
+            enabled = panorama_ha_state_result.find("./result/enabled")
+
+            if enabled == "yes":
+                # Only associate Firewalls with the active Panorama instance
+                state = find_text_in_element(panorama_ha_state_result, "./result/group/local-info/state")
+                if "active" in state:
+                    # TODO: Work out how to get the Panorama peer serial..
+                    self.ha_active_devices[device.serial] = "peer serial not implemented here.."
+                    self.get_all_child_firewalls(device)
+                    return
+            else:
+                self.get_all_child_firewalls(device)
+
+            # This is a bit of a hack - if no ha, treat it as active
+            self.ha_active_devices[device.serial] = "STANDALONE"
+            self.panorama_objects[device.serial] = device
+
+            return
+
+        elif type(device) is Firewall:
+            self.firewall_objects[device.serial] = device
+            return
+
+        raise TypeError(f"{type(device)} is not valid as a topology object.")
+
+    def top_level_devices(self) -> Generator[Union[Firewall, Panorama], None, None]:
```

Comment:
This function is kind of an interface - Firewall And Panorama objects both act very similar so it's purpose is to yield any top level device regadless of type.

Reply:
so maybe you can keep that function and split the code inside into two functions as mentioned above. WDYT?

Reply:
I split this up slightly - there's no benefit i can see for having two top_level_devices functions as the logic is very simple (it's literally "is there panorama devices? ok, then return those. Otherwise, return the firewall devices". Having top_level_firewall_devices would never be used.

In the future we may expand this logic to include device-groups and templates, where we would have different functions for things like `get_devices_by_top_level_device_group(device_group)` but that's not required now.